Update save_on_cpu and checkpointing to work with functorch wrapped tensors

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89166
* #85849
* #88357

Design doc: https://docs.google.com/document/d/1OX5__xKsZP-natgEnsRrK4gfD0wSN9WS6j4kekfn-IA/edit

This approach saves the inner-most tensor. As we unwrap, we also append to a list of rewrap functions that capture the necessary information to restore the metadata on the original wrappers. This PR tries to do most things in Python, but there are probably some APIs that could exist (or maybe already exist) that could simplify this PR.

- This PR does very weird things to stash autograd metadata:
  - The rewrap function needs to capture autograd metadata so that the higher order graphs don't get disconnected, we reuse TensorWrapper to do this, but in a way that is careful not to save the original TensorWrapper's data
  - During packing, we do a clone on the original TensorWrapper, then replace the value_ with an empty tensor, so this new dataless TensorWrapper gets captured instead by rewrap fn
  - During unpacking, when we run the rewrap fn, we just replace the value_ with the value we desire (this could either be the 
    recomputed value or value that was previously offloaded)
  - The API exposed to replace value_ is set_data!
- There doesn't seem to be a reliable way to uniquely identify a tensor since id() gets reused, using data_ptr helps but it is 
  also not enough sometimes. In this PR, I'm also using the first element of the Tensor to get a test to pass.

Unanswered questions:
- Why did we need to enable grad mode while packing (where was it disabled)

Other prototypes:
- https://github.com/pytorch/pytorch/pull/89159 (alternate approach that saves the outer-most tensor instead and unwraps the necessary number of layers during unpack - the issue is that we cannot tell when we are saving the outer-most tensor)
- https://github.com/pytorch/pytorch/pull/88976 (same approach as this PR, but in cpp, unfinished)

TODO:
- verify in tests that we are actually saving the correct amount of tensors
- try a non-zero bdim
- make that assert more precise

soulitzer committed:
Update save_on_cpu to dedupe tensors and work with functorch wrapped tensors

[ghstack-poisoned]

soulitzer committed:
Update on "Update save_on_cpu to dedupe tensors and work with functorch wrapped tensors"


This approach saves the inner-most tensor. As we unwrap, we also append to a list of rewrap functions that capture the necessary information to restore the metadata on the original wrappers.

Other prototypes:
- https://github.com/pytorch/pytorch/pull/89159 (alternate approach that saves the outer-most tensor instead and unwraps the necessary number of layers during unpack - the issue is that we cannot tell when we are saving the outer-most tensor)
- https://github.com/pytorch/pytorch/pull/88976 (same approach as this PR, but in cpp, unfinished)

TODO:
- verify in tests that we are actually saving the correct amount of tensors
- try a non-zero bdim
- make that assert more precise


[ghstack-poisoned]

soulitzer committed:
Update on "Update save_on_cpu to dedupe tensors and work with functorch wrapped tensors"


This approach saves the inner-most tensor. As we unwrap, we also append to a list of rewrap functions that capture the necessary information to restore the metadata on the original wrappers.

Other prototypes:
- https://github.com/pytorch/pytorch/pull/89159 (alternate approach that saves the outer-most tensor instead and unwraps the necessary number of layers during unpack - the issue is that we cannot tell when we are saving the outer-most tensor)
- https://github.com/pytorch/pytorch/pull/88976 (same approach as this PR, but in cpp, unfinished)

TODO:
- verify in tests that we are actually saving the correct amount of tensors
- try a non-zero bdim
- make that assert more precise


[ghstack-poisoned]

soulitzer commented on torch/autograd/graph.py:
```diff
@@ -127,24 +226,82 @@ class save_on_cpu(saved_tensors_hooks):
+                        inner.size(),
+                        dtype=inner.dtype,
+                        layout=inner.layout,
+                        pin_memory=(torch.cuda.is_available() and not inner.is_sparse))
```
inner -> unwrapped

soulitzer commented on torch/autograd/graph.py:
```diff
@@ -127,24 +226,82 @@ class save_on_cpu(saved_tensors_hooks):
+            # If saved tensor logic properly detached, we shouldn't have to unwrap and rewrap
+            # here at all. The rewrapping here is implicit due to lifting.
+            assert torch._C._functorch.maybe_get_level(ret) > original_level or original_level == -1, (
+                "lifting should happen unless we've left transforms entirely")
```
Not sure this is true actually, if we consider the case;

grad(
   forward
        grad(
               forward    (tensor packed here)
               backward (or here)
   backward             (but unpacked here)

Lifting shouldn't happen, but I haven't seen the assert triggered.

soulitzer commented:
@zou3519  In case you started to read this, I plan to rewrite the checkpointing logic actually, which actually may turn out to be a lot simpler than save_on_cpu (which I had mostly been thinking about before this).

From Alban's comments offline:
- the checkpoint application may be able to be simplified (unlike save_on_cpu, we don't really need to worry about recomputing both grad(t) and t), this should get rid of most of the unwrapping + wrapping logic.
- need to consider the case where we checkpoint a function that calls backward inside (just need to early stop when when the size of the storage reaches the size of what we expect to be saved) or else it would infinite loop
- agrees that reenabling functorch modes if backward is done later should be done, and should be straightforward

zou3519 commented:
> need to consider the case where we checkpoint a function that calls backward inside (just need to early stop when when the size of the storage reaches the size of what we expect to be saved) or else it would infinite loop

For this case, let's say we're doing
```py
def f(x):
  y = x.sin()
  z = y.sin()
  return z

def g(x):
  return checkpoint(grad_1(f), x)

grad_0(g)(x)
```
grad_1 executes something like:
```
y = x.sin()
z = y.sin()
# --- 
recomputed_y = x.sin()
gy = recomputed_y.cos()
gx = gy * x.cos()
```
If we "early stop checkpointing", what happens in grad_0? Do gy, gx, etc get recomputed, or are they saved?

soulitzer committed:
Update on "Update save_on_cpu and checkpointing to work with functorch wrapped tensors"


Design doc: https://docs.google.com/document/d/1OX5__xKsZP-natgEnsRrK4gfD0wSN9WS6j4kekfn-IA/edit

This approach saves the inner-most tensor. As we unwrap, we also append to a list of rewrap functions that capture the necessary information to restore the metadata on the original wrappers. This PR tries to do most things in Python, but there are probably some APIs that could exist (or maybe already exist) that could simplify this PR.

- This PR does very weird things to stash autograd metadata:
  - The rewrap function needs to capture autograd metadata so that the higher order graphs don't get disconnected, we reuse TensorWrapper to do this, but in a way that is careful not to save the original TensorWrapper's data
  - During packing, we do a clone on the original TensorWrapper, then replace the value_ with an empty tensor, so this new dataless TensorWrapper gets captured instead by rewrap fn
  - During unpacking, when we run the rewrap fn, we just replace the value_ with the value we desire (this could either be the 
    recomputed value or value that was previously offloaded)
  - The API exposed to replace value_ is set_data!
- There doesn't seem to be a reliable way to uniquely identify a tensor since id() gets reused, using data_ptr helps but it is 
  also not enough sometimes. In this PR, I'm also using the first element of the Tensor to get a test to pass.

Unanswered questions:
- Why did we need to enable grad mode while packing (where was it disabled)

Other prototypes:
- https://github.com/pytorch/pytorch/pull/89159 (alternate approach that saves the outer-most tensor instead and unwraps the necessary number of layers during unpack - the issue is that we cannot tell when we are saving the outer-most tensor)
- https://github.com/pytorch/pytorch/pull/88976 (same approach as this PR, but in cpp, unfinished)

TODO:
- verify in tests that we are actually saving the correct amount of tensors
- try a non-zero bdim
- make that assert more precise


[ghstack-poisoned]
